exporters/prometheus/src/exporter_utils.cc

```diff
@@ -167,6 +193,31 @@ std::string PrometheusExporterUtils::SanitizeNames(std::string name)
   return name;
 }
 
+void PrometheusExporterUtils::AddPrometheusLabel(
+    std::string name,
+    std::string value,
+    std::vector<::prometheus::ClientMetric::Label> *labels)
+{
+  prometheus_client::ClientMetric::Label prometheus_label;
+  prometheus_label.name  = std::move(name);
+  prometheus_label.value = std::move(value);
+  labels->emplace_back(std::move(prometheus_label));
+}
+
+bool PrometheusExporterUtils::ShouldIgnoreResourceAttribute(const std::string &name)
+{
+  static std::unordered_set<std::string> ignores{
```

Comment:
Why are you ignoring these resource attributes?

Edit: ah, I see your reasoning below.  I think you can remove this filtering.

Reply:
According to https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#resource-attributes , some attributes MUST NOT be added as metric attributes.
And ignoring these attributes is also used to avoid duplicate `job` and `instance`.

Reply:
This is the relevant section for SDK exporters:

> In SDK Prometheus (pull) exporters, resource attributes SHOULD be converted to a single [target_info metric](https://github.com/OpenObservability/OpenMetrics/blob/main/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems); otherwise, they MUST be dropped, and MUST NOT be attached as labels to other metric families. The target_info metric MUST be an info-typed metric whose labels MUST include the resource attributes, and MUST NOT include any other labels. There MUST be at most one target_info metric exposed on an SDK Prometheus endpoint.

Where do you see that some attributes MUST NOT be added as metric attributes?

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/prometheus_and_openmetrics.md#resource-attributes

<blockquote>
<!--StartFragment--><p dir="auto">The following attributes SHOULD be associated with scraped metrics as resource
attributes, and MUST NOT be added as metric attributes:</p>


OTLP Resource Attribute | Description
-- | --
server.address | The <host> portion of the target's URL that was scraped
server.port | The <port> portion of the target's URL that was scraped
url.scheme | http or https

<!--EndFragment-->
</blockquote>

Thanks, I may have some misunderstanding before. I remove the ignores in metric attributes, the rest ignores in resource attributes are jus used to avoid duplicate `job` and `instance`. And kee the same behaviour as https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/collector.go#L302 .

Reply:
Ah, that is in the Prometheus -> OTLP section.  This exporter is OTLP -> Prometheus.

Reply:
I remove the ignores in metric attributes, the rest ignores in resource attributes are just used to avoid duplicate `job` and `instance` labels. The behaviour is just like [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/prometheusexporter/collector.go#L302) now .

Reply:
Please do not ignore any labels.  The otel-collector-contrib exporter is _not_ a good comparison point, as it follows a different specification.  I would recommend comparing to the [otel-go](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/prometheus/exporter.go#L165) implementation, the [otel-java](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/prometheus/src/main/java/io/opentelemetry/exporter/prometheus/Serializer.java#L246) implementation, or the [otel-rust](https://github.com/open-telemetry/opentelemetry-rust/blob/main/opentelemetry-prometheus/src/lib.rs#L300) implementation.  Those languages are all fully complaint with the prometheus specification for SDKs.

Reply:
This PR do not ignore any label now, The ignore list is used to avoid duplicated adding `instance` and `job` label.[otel-go](https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/prometheus/exporter.go#L294) also use `map[string][]string` to merge keys and avoid duplicate, but prometheus-cpp use `vector`(just like array list), which can not ensure `instance` and `job` only be added once.

Reply:
You do not need to worry about duplicate job or instance labels, since this exporter does not add `target` or `instance` labels.  If a user did add those labels as resource attributes, they would be overwritten by the prometheus server at scrape time unless the prometheus server is configured with `honor_labels=true`. By not filtering it here, we give the prometheus operator the decision of whether to keep them or not.

Duplicate label keys are handled by changes from https://github.com/open-telemetry/opentelemetry-cpp/pull/2326, so you do not need to worry about that in this PR.  This exporter takes a different (and more efficient) approach compared to otel-go, which is similar to what otel-java does: https://github.com/open-telemetry/opentelemetry-java/pull/5717.  It relies on attributes being sorted to avoid creating a map[string][]string.

Please, please do not ignore any resource attributes.

Reply:
OK, all ignores are removed now.